src/main/java/vct/col/rewrite/Triggers.scala

```diff
@@ -0,0 +1,146 @@
+      } else {
+        (args.map(collectPatterns).foldLeft(Set[ASTNode]())(_ ++ _._1), false)
+      }
+    case Dereference(obj, field) =>
+      collectPatterns(obj) match {
+        case (pats, false) =>
+          (pats, false)
+        case (pats, true) =>
+          (pats + node, true)
+      }
+    case StructValue(_, _, xs) =>
+      (xs.map(collectPatterns).foldLeft(Set[ASTNode]())(_ ++ _._1), false)
+    case _ =>
+      Abort("%s", node)
+      ???
+  }
+
+  def mentions(node: ASTNode): Set[String] = node match {
+    case NameExpression(name, _, _) => Set(name)
+    case ConstantExpression(_) => Set()
+    case MethodInvokation(_, _, _, args) => args.map(mentions).foldLeft(Set[String]())(_ ++ _)
+    case OperatorExpression(_, args) => args.map(mentions).foldLeft(Set[String]())(_ ++ _)
+    case Dereference(obj, _) => mentions(obj)
+  }
+
+  def contains(container: ASTNode, element: ASTNode): Boolean =
+    container == element || (container match {
+      case _: NameExpression | _: ConstantExpression => false
+      case MethodInvokation(_, _, _, args) => args.exists(contains(_, element))
+      case OperatorExpression(_, args) => args.exists(contains(_, element))
+      case Dereference(obj, _) => contains(obj, element)
+    })
+
+  def powerset[T](x: Set[T]): Seq[Set[T]] = {
+    if(x.isEmpty) {
+      Seq(Set())
+    } else {
+      val tail = powerset(x.tail)
+      tail ++ tail.map(_ + x.head)
+    }
+  }
+
+  def computeTriggers(names: Set[String], node: ASTNode): Seq[Set[ASTNode]] = {
+    val patterns = collectPatterns(node)._1.filterNot(mentions(_).intersect(names).isEmpty)
+    val patternsNoDirectChild = patterns.filter {
+      case Dereference(obj, _) => !patterns.contains(obj)
+      case OperatorExpression(Size, List(xs)) => !patterns.contains(xs)
+      case MethodInvokation(_, _, "alen", Seq(arr)) => !patterns.contains(arr)
+      case _ => true
+    }
+
+    var triggers = powerset(patternsNoDirectChild)
+    triggers = triggers.filter(_.foldLeft(Set[String]())(_ ++ mentions(_)).intersect(names) == names)
+    triggers = triggers.filter(set => set.forall(big => set.forall(small => big == small || !contains(big, small))))
+    triggers
+  }
+
+  def tryComputeTrigger(decls: Array[DeclarationStatement], cond: ASTNode, body: ASTNode): Option[Set[Set[ASTNode]]] = {
+    val names = decls.map(_.name).toSet
+
+    computeTriggers(names, body) match {
+      case Seq() => None
+      case Seq(set) => Some(Set(set))
+      case triggers => body match {
+        case op@OperatorExpression(EQ | NEQ | LT | GT | LTE | GTE, List(left, right)) =>
+          val leftTriggers = triggers.filter(_.forall(contains(left, _)))
+          val rightTriggers = triggers.filter(_.forall(contains(right, _)))
+          if(leftTriggers.size == 1 && rightTriggers.size == 1) {
+            // Allow relational operators at the top if both sides generate exactly one trigger
```

Comment:
Yes this is possible, but we should be extremely conservative in the triggers we generate by default. This could easily lead to accidental matching loops and whatnot. In this case it's important to recognize that the set of triggers would be `{leftTrigger} {rightTrigger}`, not `{leftTrigger, rightTrigger}`.